File: sdks/go/transactions.go
package clawdb

import (
	"context"
	"encoding/json"
	"fmt"
)

// TxClient provides transactional memory operations.
type TxClient struct {
	*httpClient
}

func newTxClient(cfg *Config, session *Session) *TxClient {
	return &TxClient{newHTTPClient(cfg, session)}
}

// Begin starts a transaction. POST /v1/tx
func (t *TxClient) Begin(ctx context.Context) (string, error) {
	out, err := t.post(ctx, "/v1/tx", nil)
	if err != nil {
		return "", err
	}
	txID, _ := out["tx_id"].(string)
	return txID, nil
}

// Remember adds a memory to a transaction. POST /v1/tx/:id/memories
func (t *TxClient) Remember(ctx context.Context, txID, content string) (string, error) {
	out, err := t.post(ctx, fmt.Sprintf("/v1/tx/%s/memories", txID), map[string]interface{}{"content": content})
	if err != nil {
		return "", err
	}
	memID, _ := out["memory_id"].(string)
	return memID, nil
}

// RememberTyped adds a typed memory to a transaction. POST /v1/tx/:id/memories/typed
func (t *TxClient) RememberTyped(ctx context.Context, txID, content, memType string, tags []string, metadata map[string]interface{}) (string, error) {
	metaJSON, _ := json.Marshal(metadata)
	out, err := t.post(ctx, fmt.Sprintf("/v1/tx/%s/memories/typed", txID), map[string]interface{}{
		"content":       content,
		"type":          memType,
		"tags":          tags,
		"metadata_json": string(metaJSON),
	})
	if err != nil {
		return "", err
	}
	memID, _ := out["memory_id"].(string)
	return memID, nil
}

// Commit commits a transaction. POST /v1/tx/:id/commit
func (t *TxClient) Commit(ctx context.Context, txID string) (bool, error) {
	out, err := t.post(ctx, fmt.Sprintf("/v1/tx/%s/commit", txID), nil)
	if err != nil {
		return false, err
	}
	committed, _ := out["committed"].(bool)
	return committed, nil
}

// Rollback rolls back a transaction. POST /v1/tx/:id/rollback
func (t *TxClient) Rollback(ctx context.Context, txID string) (bool, error) {
	out, err := t.post(ctx, fmt.Sprintf("/v1/tx/%s/rollback", txID), nil)
	if err != nil {
		return false, err
	}
	rolledBack, _ := out["rolled_back"].(bool)
	return rolledBack, nil
}

File: sdks/python/clawdb/transactions.py
"""Transactional memory client for ClawDB."""
from __future__ import annotations

import json
from typing import Any

from clawdb._transport import make_metadata, with_retry
from clawdb.errors import ClawDBError
from clawdb.models import TxInfo


class TxClient:
    def __init__(self, stub: Any, session: Any, api_key: str = "") -> None:
        self._stub = stub
        self._session = session
        self._api_key = api_key

    def _meta(self) -> list[tuple[str, str]]:
        return make_metadata(self._session.token, self._api_key)

    def begin(self) -> TxInfo:
        def _call() -> TxInfo:
            try:
                resp = self._stub.BeginTx({}, metadata=self._meta())
                return TxInfo.model_validate(_to_dict(resp))
            except Exception as exc:
                raise ClawDBError.from_grpc_error(exc) from exc

        return with_retry(_call)

    def remember(self, tx_id: str, content: str) -> str:
        def _call() -> str:
            try:
                resp = self._stub.TxRemember({"tx_id": tx_id, "content": content}, metadata=self._meta())
                return str(resp.memory_id) if hasattr(resp, "memory_id") else _to_dict(resp).get("memory_id", "")
            except Exception as exc:
                raise ClawDBError.from_grpc_error(exc) from exc

        return with_retry(_call)

    def remember_typed(self, tx_id: str, content: str, *, type: str = "context", tags: list[str] | None = None, metadata: dict[str, Any] | None = None) -> str:
        def _call() -> str:
            try:
                resp = self._stub.TxRememberTyped(
                    {
                        "tx_id": tx_id,
                        "content": content,
                        "type": type,
                        "tags": tags or [],
                        "metadata_json": json.dumps(metadata or {}),
                    },
                    metadata=self._meta(),
                )
                return str(resp.memory_id) if hasattr(resp, "memory_id") else _to_dict(resp).get("memory_id", "")
            except Exception as exc:
                raise ClawDBError.from_grpc_error(exc) from exc

        return with_retry(_call)

    def commit(self, tx_id: str) -> bool:
        def _call() -> bool:
            try:
                resp = self._stub.CommitTx({"tx_id": tx_id}, metadata=self._meta())
                return bool(resp.committed) if hasattr(resp, "committed") else bool(_to_dict(resp).get("committed"))
            except Exception as exc:
                raise ClawDBError.from_grpc_error(exc) from exc

        return with_retry(_call)

    def rollback(self, tx_id: str) -> bool:
        def _call() -> bool:
            try:
                resp = self._stub.RollbackTx({"tx_id": tx_id}, metadata=self._meta())
                return bool(resp.rolled_back) if hasattr(resp, "rolled_back") else bool(_to_dict(resp).get("rolled_back"))
            except Exception as exc:
                raise ClawDBError.from_grpc_error(exc) from exc

        return with_retry(_call)


def _to_dict(msg: Any) -> dict[str, Any]:
    if isinstance(msg, dict):
        return msg
    try:
        from google.protobuf.json_format import MessageToDict
        return MessageToDict(msg, preserving_proto_field_name=True)
    except Exception:
        return {}
